text/0000-app-namespacing-semantic-convention.md

```diff
@@ -0,0 +1,61 @@
+* `app.cart.items.count`
+* `app.shipping.cost.total`
+
+The more specific the namespace, the less likely specific names can collide in your app.
+
+DO use a period (`.`) to separate sub-namespaces.
+DO NOT use characters other than a period to separate sub-namespaces.
+DO NOT mix different characters to separate sub-namespaces.
+
+## Internal details
+
+This would be accomplished by defining the `app.` prefix as reserved for end-user apps in the specification. Additionally, a specification entry in relevant places (trace, metrics, logs, and resource semantic conventions).
+
+Additionally, official documentation would be added:
+
+* Code samples in documentation updated to follow the convention as needed
+* Subsections in language SDK docs that explain how to add attributes will explain or link to this convention
+* Dedicated section in documentation about best practices for organizing data can be written, linked to, and link to the specification
+
+## Trade-offs and mitigations
+
+The biggest negative to introducing this convention now is that anyone using a _different_ convention will be considered out of step with best practices. For them to follow best practices, they would need to rewrite their instrumentation names, which would likely break or complicate their analyses for a period of time. To mitigate this, several analysis tools offer ways to coalesce two field names into one, but this kind of functionality is not guaranteed to be present for everyone.
+
+Another negative is that some developers may not like the idea of defining a convention for their own systems. It's not uncommon for developers to dislike standard coding conventions, for example, and we could expect to see a similar kind of backlash around defining this kind of convetions. There isn't a really a way to mitigate this kind of feedback; however, there is precedent that most developers tend to like conventions to follow rather than having to define their own all of the time. And so it's likely that most developers would appreciate following this convention.
+
+Finally, like any standard convention, it's not much use if it's not documented. The proposed additions to documentation (even just having the right link the right places) are nontrivial. One potential way to mitigate this could be to use tooling to somehow suggest or enforce this convention (e.g., Analyzers in C#). However, that is a much bigger lift than defining a spec convention, and not all languages are guaranteed to have that kind of tooling available.
+
+## Prior art and alternatives
+
+Namespacing attributes is a documented best practice for [some vendors](https://docs.honeycomb.io/getting-data-in/data-best-practices/#namespace-custom-fields). These practices emerged before OpenTelemetry, but are equally applicable to OpenTelemetry.
+
+This convention is also being used by the [OpenTelemetry Community Demo](https://github.com/open-telemetry/opentelemetry-demo/blob/main/src/checkoutservice/main.go#L265-L267).
```

Comment:
It would be risky to use attributes names not part of the spec.  IMHO, as a best practice, the community demo should also not be using the app namespace as it's not in the spec.

Reply:
@scheler, maybe worth raising this as an issue with the community demo?

Reply:
Best to raise it there. They were intentional about doing it but I don't know exactly what the decision process was. FWIW I think the demo should establish conventions for things like attribute naming even if they're not in the spec.